### PR TITLE
Start generating new shared code

### PR DESCRIPTION
**Pull request by richfitz**
Generates code compatible with https://github.com/mrc-ide/dust/pull/168

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -36,5 +36,5 @@ Suggests:
 RoxygenNote: 7.1.1
 VignetteBuilder: knitr
 Remotes:
-    mrc-ide/dust,
+    mrc-ide/dust@i116-shared-gpu,
```
Comment (johnlees) [inline DESCRIPTION:39]:
```suggestion
    mrc-ide/dust,
```

File: R/generate_dust.R [hunk 2]
```diff
@@ -624,20 +626,49 @@ generate_dust_gpu <- function(eqs, dat, rewrite) {
...
+  ## Always sort scalars first; this is required later
+  i <- order(shared_rank)
+  shared <- shared[i]
+  shared_rank <- shared_rank[i]
+  shared_type <- vcapply(dat$data$elements[shared], "[[", "storage_type")
+
+  dat$gpu <- list(
+    internal = list(
+      int = intersect(internal, names_if(internal_type == "int")),
+      real = intersect(internal, names_if(internal_type == "double"))),
```
Comment (johnlees) [inline R/generate_dust.R:663]: They are doubles here, and become real_t later?

Comment (johnlees) [reply R/generate_dust.R:663]: (looks like it, but just 'double'-checking)

**Review by johnlees — APPROVED** [review]

File: R/generate_dust.R (hunk 2 above)
Comment (richfitz) [reply R/generate_dust.R:663]: yeah, in the underling odin IR we have types "bool", "int", "double" (https://github.com/mrc-ide/odin/blob/master/inst/schema.json#L41-L42) but in dust we reat things that are "double" as "real_t"

**Pull request merged**